### PR TITLE
TS TODOs : makes reconnect-util generic, provides type annotations

### PR DESCRIPTION
**Pull request by manavmehta**
**What's this PR do?**
What does this PR do:
    Worked on TypeScript TODOs
	Can pass class into reconnect-util from main.ts (generic class)
	CurrentWindow now has a .send() property
	We now use ProxyRule for the proxy property
	Type annotated some variables

**You have tested this PR on:**
  - [✓ ] macOS (Catalina 10.15.2)

Zulip App v 4.0.3

### CODE REVIEW

File: app/renderer/js/components/handle-external-link.ts [hunk 1]
```diff
@@ -75,7 +76,7 @@ function handleExternalLink(this: any, event: any): void {
 						body: 'Download failed'
 					});
 				} else {
-					this.$el.downloadURL(url);
+					this.downloadURL(url);
```
Comment (vsvipul) [inline app/renderer/js/components/handle-external-link.ts]: Why was this changed? This will also change the functioning, i guess.

File: app/renderer/js/components/handle-external-link.ts [hunk 2]
```diff
@@ -1,15 +1,16 @@
-import { ipcRenderer, remote } from 'electron';
+import { ipcRenderer, remote, NewWindowEvent } from 'electron';
```
Comment (vsvipul) [inline app/renderer/js/components/handle-external-link.ts]: Use Electron.NewWindowEvent instead of importing.

File: app/renderer/js/notification/helpers.ts [hunk 3]
```diff
@@ -82,7 +82,7 @@ const webContentsId = webContents.id;
 export function focusCurrentServer(): void {
 	// TODO: TypeScript: currentWindow of type BrowserWindow doesn't
 	// have a .send() property per typescript.
-	(currentWindow as any).send('focus-webview-with-id', webContentsId);
+	currentWindow.webContents.send('focus-webview-with-id', webContentsId);
```
Comment (vsvipul) [inline app/renderer/js/notification/helpers.ts:83]: Why webcontents is added ? This is wrong.

File: app/renderer/js/pages/preference/general-section.ts [hunk 4]
```diff
@@ -288,7 +288,7 @@ class GeneralSection extends BaseSection {
 				this.updateSilentOption();
 				// TODO: TypeScript: currentWindow of type BrowserWindow doesn't
 				// have a .send() property per typescript.
-				(currentBrowserWindow as any).send('toggle-silent', newValue);
+				currentBrowserWindow.webContents.send('toggle-silent', newValue);
```
Comment (vsvipul) [inline app/renderer/js/pages/preference/general-section.ts:293]: Same here.

File: app/renderer/js/utils/link-util.ts [hunk 5]
```diff
@@ -22,12 +22,12 @@ class LinkUtil {
 	}
 
 	isInternal(currentUrl: string, newUrl: string): IsInternalResponse {
-		const currentDomain = wurl('hostname', currentUrl);
-		const newDomain = wurl('hostname', newUrl);
+		const currentDomain: string = wurl('hostname', currentUrl);
```
Comment (vsvipul) [inline app/renderer/js/utils/link-util.ts]: Explicitly mentioning type is not required anywhere. Same for all 4 changes below.

**Comment by vsvipul** [comment]
Left some comments. Also changes to package-lock.json are not required to be added to the PR anywhere. Remove those. Also, TODO messages also need to be removed when fixing that todo. I would suggest fixing all TODOs in a single commit, as all of them are minor changes. Also better commit messages and PR names and description will help.

File: app/renderer/js/utils/link-util.ts (hunk 5 above)
Comment (manavmehta) [reply app/renderer/js/utils/link-util.ts]: A doubt here actually! When (in VS Code), I hover to these vars (currentDomain and newDomain only), they say :any. So isn't an explicit mention required to avoid the type 'any' ?

File: app/renderer/js/notification/helpers.ts (hunk 3 above)
Comment (manavmehta) [reply app/renderer/js/notification/helpers.ts:83]: BrowserWindow as per electronjs docs, doesn't have any .send() property. It is a property used as BrowserWindow.webContents.send().
That's why I did this.

Comment (vsvipul) [reply app/renderer/js/notification/helpers.ts:83]: Test the functionality. If it works this way, then it is fine.

File: app/renderer/js/utils/link-util.ts (hunk 5 above)
Comment (vsvipul) [reply app/renderer/js/utils/link-util.ts]: No, not required.

File: app/renderer/js/components/handle-external-link.ts (hunk 2 above)
Comment (manavmehta) [reply app/renderer/js/components/handle-external-link.ts]: Done

File: app/renderer/js/notification/helpers.ts (hunk 3 above)
Comment (manavmehta) [reply app/renderer/js/notification/helpers.ts:83]: Tested. Works Fine.

**Comment by manavmehta** [comment]
Hey @vsvipul I have fixed it according to your review. can you please re-review and merge it ?
cc: @akashnimare

**Comment by abhigyank** [comment]
The completed ToDos look fine to me for this PR. Commit message can be improved while merging.

**Comment by manavmehta** [comment]
@abhigyank I'll keep that in mind and henceforth provide better commit messages

**Review by vsvipul — APPROVED** [review]
LGTM!

File: app/renderer/js/components/handle-external-link.ts [hunk 6]
```diff
@@ -9,7 +9,7 @@ const { shell, app } = remote;
 const dingSound = new Audio('../resources/sounds/ding.ogg');
 
 // TODO: TypeScript - Figure out a way to pass correct type here.
-function handleExternalLink(this: any, event: any): void {
+function handleExternalLink(this: any, event: Electron.NewWindowEvent): void {
```
Comment (andersk) [inline app/renderer/js/components/handle-external-link.ts]: Should be `this: WebView` (with `import type WebView from './webview'`).

File: app/renderer/js/utils/reconnect-util.ts [hunk 7]
```diff
@@ -11,11 +11,8 @@ const logger = new Logger({
 	timestamp: true
 });
 
-class ReconnectUtil {
-	// TODO: TypeScript - Figure out how to annotate webview
-	// it should be WebView; maybe make it a generic so we can
-	// pass the class from main.ts
-	webview: any;
+class ReconnectUtil<ClassName> {
+	webview: ClassName ;
```
Comment (andersk) [inline app/renderer/js/utils/reconnect-util.ts]: This generic doesn’t make sense because we do `webview.props.url`. TypeScript doesn’t catch that only because you forgot to change the `webview: any` parameter of the constructor. We should just use `webview: WebView` in both places.

File: app/renderer/js/utils/request-util.ts [hunk 8]
```diff
@@ -17,11 +17,14 @@ const logger = new Logger({
 
 let instance: null | RequestUtil = null;
 
-// TODO: TypeScript - Use ProxyRule for the proxy property
-// we can do this now since we use export = ProxyUtil syntax
+interface ProxyRule {
+	hostname?: string;
+	port?: number;
+}
```
Comment (andersk) [inline app/renderer/js/utils/request-util.ts]: This is already defined in `proxy-util.ts`; the comment is suggesting we should import it, not redefine it.

**Comment by andersk** [comment]
Updated as suggested and merged. Thanks @manavmehta!

**Comment by manavmehta** [comment]
Hey @andersk I had these changes in mind but I've been encountering some warnings so avoided them purposefully. I've commented all the warnings above replying to each of your comment.
Are these to be ignored ?

File: app/renderer/js/components/handle-external-link.ts (hunk 6 above)
Comment (manavmehta) [reply app/renderer/js/components/handle-external-link.ts]: Cannot find namespace 'WebView'   .ts(2503)

File: app/renderer/js/utils/reconnect-util.ts (hunk 7 above)
Comment (manavmehta) [reply app/renderer/js/utils/reconnect-util.ts]: Did I get the TODO wrong as you're correct that `webview.props.url` even after changing the constructor to `webview: ClassName`

P.S. It was asked to make it generic

Comment (andersk) [reply app/renderer/js/utils/reconnect-util.ts]: You read the TODO correctly, but the TODO did not make sense, and we should do things that make sense. :slightly_smiling_face:

File: app/renderer/js/components/handle-external-link.ts (hunk 6 above)
Comment (andersk) [reply app/renderer/js/components/handle-external-link.ts]: It works fine on `master`. Did you forget to run `npm install` to follow the upgrades we did to TypeScript and other dependencies?

Comment (manavmehta) [reply app/renderer/js/components/handle-external-link.ts]: solved
